File: deepwiki_to_md/chat/deep_research.py
"""
Deep Research Mode for DeepWiki to Markdown Converter

This module provides functionality for conducting deep research using the DeepWiki chat.
"""

import logging
import time
import os
from typing import Dict, List, Optional, Any, Union
import json

from deepwiki_to_md.models.config import Config
from deepwiki_to_md.chat.chat_interface import ChatInterface
from deepwiki_to_md.utils.file_io import save_markdown_file, save_json_file, save_yaml_file
from deepwiki_to_md.converters.md_to_yaml import convert_markdown_to_yaml, yaml_to_string

logger = logging.getLogger(__name__)


class DeepResearchMode:
    """
    Deep Research Mode for the DeepWiki chat.
    
    This class provides methods for conducting deep research using the DeepWiki chat,
    including multi-turn conversations and saving results in various formats.
    """

    def __init__(self, config: Config):
        """
        Initialize the deep research mode.
        
        Args:
            config (Config): Configuration object
        """
        self.config = config
        self.chat = None
        self.conversation_history = []
        self.research_results = []

        logger.debug("Initialized DeepResearchMode")

    def __enter__(self):
        """
        Enter context manager.
        
        Returns:
            DeepResearchMode: Self
        """
        self.start_session()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """
        Exit context manager.
        
        Args:
            exc_type: Exception type
            exc_val: Exception value
            exc_tb: Exception traceback
        """
        self.end_session()

    def start_session(self) -> bool:
        """
        Start a deep research session.
        
        Returns:
            bool: True if session started successfully, False otherwise
        """
        try:
            # Initialize chat interface
            self.chat = ChatInterface(self.config)
            success = self.chat.connect()

            if success:
                logger.info("Started deep research session")

                # Send initial system message if configured
                system_prompt = self.config.get('system_prompt')
                if system_prompt:
                    self._add_to_history("system", system_prompt)
                    logger.debug(f"Set system prompt: {system_prompt}")

            return success

        except Exception as e:
            logger.error(f"Error starting deep research session: {str(e)}")
            return False

    def end_session(self) -> None:
        """
        End the deep research session.
        """
        try:
            if self.chat:
                self.chat.disconnect()
                self.chat = None

            logger.info("Ended deep research session")

        except Exception as e:
            logger.error(f"Error ending deep research session: {str(e)}")

    def ask_question(self, question: str, timeout: int = 120) -> Optional[str]:
        """
        Ask a question in the deep research session.
        
        Args:
            question (str): Question to ask
            timeout (int): Maximum time to wait for a response in seconds
            
        Returns:
            Optional[str]: Response text or None if no response received
        """
        if not self.chat:
            logger.error("No active chat session")
            return None

        try:
            # Send the question
            logger.info(f"Asking question: {question}")
            self._add_to_history("user", question)

            if not self.chat.send_message(question):
                logger.error("Failed to send question")
                return None

            # Get the response
            response = self.chat.get_response(timeout)

            if response:
                logger.info("Received response")
                self._add_to_history("assistant", response)
                self.research_results.append({
                    "question": question,
                    "response": response
                })
            else:
                logger.warning("No response received")

            return response

        except Exception as e:
            logger.error(f"Error asking question: {str(e)}")
            return None

    def conduct_research(self, questions: List[str], timeout: int = 120) -> List[Dict[str, str]]:
        """
        Conduct research by asking a series of questions.
        
        Args:
            questions (List[str]): List of questions to ask
            timeout (int): Maximum time to wait for each response in seconds
            
        Returns:
            List[Dict[str, str]]: List of question-response pairs
        """
        results = []

        for i, question in enumerate(questions):
            logger.info(f"Research question {i + 1}/{len(questions)}")

            response = self.ask_question(question, timeout)

            if response:
                results.append({
                    "question": question,
                    "response": response
                })

            # Wait between questions to avoid rate limiting
            if i < len(questions) - 1:
                time.sleep(2)

        return results

    def save_results(self, output_dir: str, formats: List[str] = ["md", "json", "yaml"]) -> Dict[str, List[str]]:
        """
        Save research results in various formats.
        
        Args:
            output_dir (str): Directory to save results in
            formats (List[str]): List of formats to save results in (md, json, yaml)
            
        Returns:
            Dict[str, List[str]]: Dictionary mapping format to list of saved file paths
        """
        saved_files = {fmt: [] for fmt in formats}

        # Create output directory if it doesn't exist
        os.makedirs(output_dir, exist_ok=True)

        # Save each result
        for i, result in enumerate(self.research_results):
            question = result["question"]
            response = result["response"]

            # Create a filename based on the question
            base_filename = self._create_filename(question, i)

            # Save in each requested format
            for fmt in formats:
                if fmt.lower() == "md":
                    # Save as Markdown
                    filename = f"{base_filename}.md"
                    filepath = os.path.join(output_dir, filename)

                    content = f"# {question}\n\n{response}"
                    if save_markdown_file(filepath, content):
                        saved_files["md"].append(filepath)

                elif fmt.lower() == "json":
                    # Save as JSON
                    filename = f"{base_filename}.json"
                    filepath = os.path.join(output_dir, filename)

                    if save_json_file(filepath, result):
                        saved_files["json"].append(filepath)

                elif fmt.lower() == "yaml":
                    # Save as YAML
                    filename = f"{base_filename}.yaml"
                    filepath = os.path.join(output_dir, filename)

                    # Convert to YAML structure
                    yaml_data = convert_markdown_to_yaml(
                        response,
                        title=question,
                        include_raw_content=True
                    )

                    # Convert to YAML string and save
                    yaml_str = yaml_to_string(yaml_data)

                    with open(filepath, 'w', encoding='utf-8') as f:
                        f.write(yaml_str)

                    saved_files["yaml"].append(filepath)

        # Save conversation history
        if self.conversation_history:
            history_file = os.path.join(output_dir, "conversation_history.json")
            save_json_file(history_file, {"history": self.conversation_history})

        return saved_files

    def _add_to_history(self, role: str, content: str) -> None:
        """
        Add a message to the conversation history.
        
        Args:
            role (str): Role of the message sender (user, assistant, system)
            content (str): Message content
        """
        self.conversation_history.append({
            "role": role,
            "content": content,
            "timestamp": time.time()
        })

    def _create_filename(self, question: str, index: int) -> str:
        """
        Create a filename based on a question.
        
        Args:
            question (str): Question to create filename from
            index (int): Index of the question
            
        Returns:
            str: Sanitized filename
        """
        # Truncate and sanitize the question for use as a filename
        sanitized = "".join(c if c.isalnum() or c in " -_" else "_" for c in question)
        sanitized = sanitized.strip()

        # Truncate if too long
        if len(sanitized) > 50:
            sanitized = sanitized[:47] + "..."

        # Add index to ensure uniqueness
        return f"{index + 1:02d}_{sanitized}"


def conduct_research(
        config: Config,
        questions: List[str],
        output_dir: Optional[str] = None,
        formats: List[str] = ["md", "json", "yaml"],
        timeout: int = 120
) -> Dict[str, Any]:
    """
    Conduct deep research and save results.
    
    Args:
        config (Config): Configuration object
        questions (List[str]): List of questions to ask
        output_dir (Optional[str]): Directory to save results in
        formats (List[str]): List of formats to save results in
        timeout (int): Maximum time to wait for each response in seconds
        
    Returns:
        Dict[str, Any]: Research results and saved file information
    """
    with DeepResearchMode(config) as research:
        # Conduct research
        results = research.conduct_research(questions, timeout)

        # Save results if output directory is provided
        saved_files = {}
        if output_dir:
            saved_files = research.save_results(output_dir, formats)

        return {
            "results": results,
            "saved_files": saved_files
        }

File: nodejs_implementation/src/chat/deep-research.js
/**
 * Deep Research Mode for DeepWiki to Markdown Converter
 * 
 * This module provides functionality for conducting deep research using the DeepWiki chat.
 */

const path = require('path');
const fs = require('fs');
const { logger } = require('../utils/logger');
const { ChatInterface } = require('./chat-interface');
const { saveMarkdownFile, saveJsonFile, saveYamlFile } = require('../utils/file-io');
const { convertMarkdownToYaml, yamlToString } = require('../converters/md-to-yaml');

/**
 * Deep Research Mode for the DeepWiki chat.
 * 
 * This class provides methods for conducting deep research using the DeepWiki chat,
 * including multi-turn conversations and saving results in various formats.
 */
class DeepResearchMode {
  /**
   * Initialize the deep research mode.
   * 
   * @param {Object} config - Configuration object
   */
  constructor(config) {
    this.config = config;
    this.chat = null;
    this.conversationHistory = [];
    this.researchResults = [];
    
    logger.debug("Initialized DeepResearchMode");
  }
  
  /**
   * Start a deep research session.
   * 
   * @returns {Promise<boolean>} True if session started successfully, False otherwise
   */
  async startSession() {
    try {
      // Initialize chat interface
      this.chat = new ChatInterface(this.config);
      const success = await this.chat.connect();
      
      if (success) {
        logger.info("Started deep research session");
        
        // Send initial system message if configured
        const systemPrompt = this.config.systemPrompt;
        if (systemPrompt) {
          this._addToHistory("system", systemPrompt);
          logger.debug(`Set system prompt: ${systemPrompt}`);
        }
      }
      
      return success;
    } catch (error) {
      logger.error(`Error starting deep research session: ${error.message}`);
      return false;
    }
  }
  
  /**
   * End the deep research session.
   * 
   * @returns {Promise<void>}
   */
  async endSession() {
    try {
      if (this.chat) {
        await this.chat.disconnect();
        this.chat = null;
      }
      
      logger.info("Ended deep research session");
    } catch (error) {
      logger.error(`Error ending deep research session: ${error.message}`);
    }
  }
  
  /**
   * Ask a question in the deep research session.
   * 
   * @param {string} question - Question to ask
   * @param {number} [timeout=120000] - Maximum time to wait for a response in milliseconds
   * @returns {Promise<string|null>} Response text or null if no response received
   */
  async askQuestion(question, timeout = 120000) {
    if (!this.chat) {
      logger.error("No active chat session");
      return null;
    }
    
    try {
      // Send the question
      logger.info(`Asking question: ${question}`);
      this._addToHistory("user", question);
      
      if (!await this.chat.sendMessage(question)) {
        logger.error("Failed to send question");
        return null;
      }
      
      // Get the response
      const response = await this.chat.getResponse(timeout);
      
      if (response) {
        logger.info("Received response");
        this._addToHistory("assistant", response);
        this.researchResults.push({
          question,
          response
        });
      } else {
        logger.warning("No response received");
      }
      
      return response;
    } catch (error) {
      logger.error(`Error asking question: ${error.message}`);
      return null;
    }
  }
  
  /**
   * Conduct research by asking a series of questions.
   * 
   * @param {Array<string>} questions - List of questions to ask
   * @param {number} [timeout=120000] - Maximum time to wait for each response in milliseconds
   * @returns {Promise<Array<Object>>} List of question-response pairs
   */
  async conductResearch(questions, timeout = 120000) {
    const results = [];
    
    // Start the session
    if (!await this.startSession()) {
      logger.error("Failed to start research session");
      return results;
    }
    
    try {
      for (let i = 0; i < questions.length; i++) {
        const question = questions[i];
        logger.info(`Research question ${i+1}/${questions.length}`);
        
        const response = await this.askQuestion(question, timeout);
        
        if (response) {
          results.push({
            question,
            response
          });
        }
        
        // Wait between questions to avoid rate limiting
        if (i < questions.length - 1) {
          await new Promise(resolve => setTimeout(resolve, 2000));
        }
      }
    } finally {
      // End the session
      await this.endSession();
    }
    
    return results;
  }
  
  /**
   * Save research results in various formats.
   * 
   * @param {string} outputDir - Directory to save results in
   * @param {Array<string>} [formats=['md', 'json', 'yaml']] - List of formats to save results in
   * @returns {Promise<Object>} Dictionary mapping format to list of saved file paths
   */
  async saveResults(outputDir, formats = ['md', 'json', 'yaml']) {
    const savedFiles = {};
    formats.forEach(fmt => { savedFiles[fmt] = []; });
    
    // Create output directory if it doesn't exist
    fs.mkdirSync(outputDir, { recursive: true });
    
    // Save each result
    for (let i = 0; i < this.researchResults.length; i++) {
      const result = this.researchResults[i];
      const question = result.question;
      const response = result.response;
      
      // Create a filename based on the question
      const baseFilename = this._createFilename(question, i);
      
      // Save in each requested format
      for (const fmt of formats) {
        if (fmt.toLowerCase() === 'md') {
          // Save as Markdown
          const filename = `${baseFilename}.md`;
          const filepath = path.join(outputDir, filename);
          
          const content = `# ${question}\n\n${response}`;
          if (saveMarkdownFile(filepath, content)) {
            savedFiles.md.push(filepath);
          }
        } else if (fmt.toLowerCase() === 'json') {
          // Save as JSON
          const filename = `${baseFilename}.json`;
          const filepath = path.join(outputDir, filename);
          
          if (saveJsonFile(filepath, result)) {
            savedFiles.json.push(filepath);
          }
        } else if (fmt.toLowerCase() === 'yaml') {
          // Save as YAML
          const filename = `${baseFilename}.yaml`;
          const filepath = path.join(outputDir, filename);
          
          // Convert to YAML structure
          const yamlData = convertMarkdownToYaml(
            response,
            question,
            null,
            { includeRawContent: true }
          );
          
          // Convert to YAML string and save
          const yamlStr = yamlToString(yamlData);
          
          fs.writeFileSync(filepath, yamlStr, 'utf8');
          savedFiles.yaml.push(filepath);
        }
      }
    }
    
    // Save conversation history
    if (this.conversationHistory.length > 0) {
      const historyFile = path.join(outputDir, "conversation_history.json");
      saveJsonFile(historyFile, { history: this.conversationHistory });
    }
    
    return savedFiles;
  }
  
  /**
   * Add a message to the conversation history.
   * 
   * @param {string} role - Role of the message sender (user, assistant, system)
   * @param {string} content - Message content
   * @private
   */
  _addToHistory(role, content) {
    this.conversationHistory.push({
      role,
      content,
      timestamp: new Date().toISOString()
    });
  }
  
  /**
   * Create a filename based on a question.
   * 
   * @param {string} question - Question to create filename from
   * @param {number} index - Index of the question
   * @returns {string} Sanitized filename
   * @private
   */
  _createFilename(question, index) {
    // Truncate and sanitize the question for use as a filename
    let sanitized = question.replace(/[<>:"/\\|?*]/g, '_');
    sanitized = sanitized.replace(/\s+/g, '_').toLowerCase();
    
    // Truncate if too long
    if (sanitized.length > 50) {
      sanitized = sanitized.substring(0, 47) + '...';
    }
    
    // Add index to ensure uniqueness
    return `${index+1}_${sanitized}`;
  }
}

/**
 * Conduct deep research and save results.
 * 
 * @param {Object} config - Configuration object
 * @param {Array<string>} questions - List of questions to ask
 * @param {string} [outputDir=null] - Directory to save results in
 * @param {Array<string>} [formats=['md', 'json', 'yaml']] - List of formats to save results in
 * @param {number} [timeout=120000] - Maximum time to wait for each response in milliseconds
 * @returns {Promise<Object>} Research results and saved file information
 */
async function conductResearch(
  config,
  questions,
  outputDir = null,
  formats = ['md', 'json', 'yaml'],
  timeout = 120000
) {
  const research = new DeepResearchMode(config);
  
  // Conduct research
  const results = await research.conductResearch(questions, timeout);
  
  // Save results if output directory is provided
  let savedFiles = {};
  if (outputDir) {
    savedFiles = await research.saveResults(outputDir, formats);
  }
  
  return {
    results,
    savedFiles
  };
}

module.exports = {
  DeepResearchMode,
  conductResearch
};